core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java

```diff
@@ -68,8 +69,20 @@ public static boolean ancestorOf(Table table, long snapshotId, long ancestorSnap
    * This method assumes that fromSnapshotId is an ancestor of toSnapshotId
    */
   public static List<Long> snapshotIdsBetween(Table table, long fromSnapshotId, long toSnapshotId) {
+    AtomicBoolean isAncestor = new AtomicBoolean(false);
```

Comment:
I think this may be clearer and not require the Atomic if we do a little refactoring and do two recursive walks,

`Precondition.require((ancestorOf(table, fromSnapshotId, toSnapshotId))`

then doing the ancestorIds lookup?

Or if we want to keep the current logic I believe we can just throw an IllegalStateException in the lambda instead of returning null or using the atomic boolean.

Reply:
I think doing two rounds of traversing is quite time consuming if we have many snapshots in a table, so I prefer not to change to that way.

> throw an IllegalStateException in the lambda instead of returning null or using the atomic boolean.

The `AtomicBoolean` here is to check if `fromSnapshotId` is not existed. We can only know if `fromSnapshotId` is existed or not after doing a full traversing, so throwing an Exception during traversing in lambda seems hard to achieve.

Reply:
you can't just have 

if (snapshotId == fromSnapshotId) throw new IllegalStateException

?